### PR TITLE
[data-client] Add Metric Counters to Data Client

### PR DESCRIPTION
**Pull request by khiemngo**
## Motivation

This PR adds metric counters to `diem-data-client` to keep track the following:
- the number of requests sent (per request type and total)
- the number of success responses (per request type and total)
- the number of error responses (per request type and total)
- the number of peer selection errors (per request type and total)
- request latencies (per request type)

### Have you read the [Contributing Guidelines on pull requests](https://github.com/diem/diem/blob/main/CONTRIBUTING.md#pull-requests)?

Yes.

## Test Plan

No test plan is required. Re-ran existing tests to make sure this PR does not fail existing tests.

## Related PRs

(If this PR adds or changes functionality, please take some time to update the docs at https://github.com/diem/diem/tree/main/developers.diem.com, and link to your PR here.)

## If targeting a release branch, please fill the below out as well

 * Justification and breaking nature (who does it affect? validators, full nodes, tooling, operators, AOS, etc.)
 * Comprehensive test results that demonstrate the fix working and not breaking existing workflows.
 * Why we must have it for V1 launch.
 * What workarounds and alternative we have if we do not push the PR.

### CODE REVIEW

File: state-sync/diem-data-client/src/diemnet/metrics.rs
Comment (khiemngo) [inline state-sync/diem-data-client/src/diemnet/metrics.rs:22]: I'm thinking of keep tracking of success responses/error responses per request type and a total of successes/errors. If you'd like to simplify by keeping track less things, please let me know.

File: state-sync/diem-data-client/src/diemnet/metrics.rs
Comment (khiemngo) [inline state-sync/diem-data-client/src/diemnet/metrics.rs:42]: I'm separating `PeerSelectionError` from the errors related to the received responses. If you want to drop this or if you want to count `PeerSelectionError` and response errors together, please let me know.

File: state-sync/diem-data-client/src/diemnet/metrics.rs
Comment (khiemngo) [inline state-sync/diem-data-client/src/diemnet/metrics.rs:62]: I'm not sure if this is a common way to do this. (Alternatively, I can have a separate counter for total (e.g., `REQUEST_COUNT`, `RESPONSE_COUNT`, etc.), but that would add more variable names and might be confusing). Please feel free to suggest another way.

File: state-sync/diem-data-client/src/diemnet/mod.rs
Comment (khiemngo) [inline state-sync/diem-data-client/src/diemnet/mod.rs:179]: Would be nice if we can keep track latency for success responses and error responses separately. But I'm not doing that yet because
- I'm not sure we need it at the moment
- I don't know an easy/simple way to do so (the current mechanism simply just stops the timer when `_timer` is dropped IIUC?)

**Comment by khiemngo** [comment]
Hi @JoshLind , I'm looking for some initial feedback on the PR. Could you please give me some feedback when you have time to take a look? Thank you! 😃

File: state-sync/diem-data-client/src/diemnet/metrics.rs
Comment (JoshLind) [inline state-sync/diem-data-client/src/diemnet/metrics.rs:9]: nit: maybe name this `TOTAL_COUNT_LABEL` and add a comment above it? e.g., the comment you added to `increment_counter` could be moved here

File: state-sync/diem-data-client/src/diemnet/mod.rs
Comment (JoshLind) [reply state-sync/diem-data-client/src/diemnet/mod.rs:179]: Yeah, I see what you mean, the timer is stopped when it's dropped. I think this should be fine for now. If we feel like we need it in the future, we can come back and figure out how to split up the timer for successes and responses. 😄 (There's ways of manually timing and then updating the counter, IIRC).

File: state-sync/diem-data-client/src/diemnet/mod.rs
Comment (JoshLind) [inline state-sync/diem-data-client/src/diemnet/mod.rs:484]: `_timer` will only be dropped at the end of this loop (when it falls out of scope), meaning it will also time `update_summary` and `update_global_summary`. To avoid this we should drop it manually after we get a response e.g., add `drop(timer)` to line 498, after we get the result?

File: state-sync/diem-data-client/src/diemnet/metrics.rs
Comment (JoshLind) [reply state-sync/diem-data-client/src/diemnet/metrics.rs:62]: Yeah, this is one way of doing it, and I think it's fine. Another way would be dynamically aggregate across all labels on the visualization side (i.e., add them all up to get the total), meaning 1 less label to deal with, but then it becomes a client-side responsibility. I think this is okay.

File: state-sync/diem-data-client/src/diemnet/metrics.rs
Comment (JoshLind) [reply state-sync/diem-data-client/src/diemnet/metrics.rs:42]: To be honest, I don't think we need this for now. If we get a peer selection error, we'll only be able to debug it using the logs. Moreover, we won't be sending any requests or receiving responses, so we'll also know that something is wrong. Perhaps let's remove this until we can see a convincing reason for it?

File: state-sync/diem-data-client/src/diemnet/metrics.rs
Comment (JoshLind) [reply state-sync/diem-data-client/src/diemnet/metrics.rs:22]: Yeah, I think this makes sense 😃

**Review by JoshLind — COMMENTED** [review]
Thanks @khiemngo! Some small nits but looks almost ready 🚀

File: state-sync/diem-data-client/src/diemnet/metrics.rs
Comment (khiemngo) [reply state-sync/diem-data-client/src/diemnet/metrics.rs]: Updated name: `pub const TOTAL_COUNT_LABEL: &str = "TOTAL_COUNT";`

File: state-sync/diem-data-client/src/diemnet/mod.rs
Comment (khiemngo) [reply state-sync/diem-data-client/src/diemnet/mod.rs:477]: Added `drop(timer)` right after receiving the response (and before `update_summary` and `update_global_summary`)

File: state-sync/diem-data-client/src/diemnet/metrics.rs
Comment (khiemngo) [reply state-sync/diem-data-client/src/diemnet/metrics.rs]: Removed the counter that tracks `PeerSelectionError`.

Yeah I agree that tracking the number of `PeerSelectionError` errors does not seem necessary/useful.

**Comment by khiemngo** [comment]
Thanks for the feedback @JoshLind . I tried to address/apply all the nits. Please let me know if there is anything else you'd like me to fix. Thank you! ❤️

File: state-sync/diem-data-client/src/diemnet/mod.rs
Comment (gregnazario) [reply state-sync/diem-data-client/src/diemnet/mod.rs:477]: I thought that `_` were also not dropped until the end of the scope (I may be wrong).  But, the whole point of it all is that `_x` or `_` doesn't bind the variable to anything and can't be used.

It's a very common pattern for anything that's needed for the scope, e.g. a lock or a timer, to do this and then for the entirety of a scope `{ ... }` be held and then dropped. More info about dropping and scope here: https://doc.rust-lang.org/book/ch15-03-drop.html

Comment (khiemngo) [reply state-sync/diem-data-client/src/diemnet/mod.rs]: Changed `_timer` to `timer`.

Thanks @gregnazario and @JoshLind for giving me more pointers to rust naming convention and dropping!

n00b question: I tried to change `_timer = start_timer(...)` to `_ = start_timer(...)` in another place where `_timer` is automatically dropped at the end of the scope. Though the compiler did not complain, this change would result in no latency being calculated since no actual timer object is bound to be able to stop? (and hence that's why we'd need to use `_timer` instead of `_`?)

Comment (gregnazario) [reply state-sync/diem-data-client/src/diemnet/mod.rs]: The compiler doesn't have an idea of what you're doing.  It's often legitimate to say `let _ = function()` just to use a "must use" return value (and I guess drop it).

So, that looks right then, you must use a name to keep it around for the scope if it's timing nothing.

Comment (khiemngo) [reply state-sync/diem-data-client/src/diemnet/mod.rs]: Yeah got it. Thanks @gregnazario! 😃

**Review by JoshLind — APPROVED** [review]
Thanks @khiemngo! LGTM! 🚀

**Comment by khiemngo** [comment]
> Thanks @khiemngo! LGTM! 🚀

Oops, sorry I just realize that I did not get the official `approval` from @JoshLind before landing. (I saw this comment and landed instead. I'll look more carefully next time 😃 ). I hope this won't be a problem...

**Comment by JoshLind** [comment]
@khiemngo, I think I did approve 🤔 Otherwise, how would you land? Either way, it's good to go 😄

**Pull request closed without merging**
